crates/prover/src/adapter/instructions.rs

```diff
@@ -0,0 +1,474 @@
+use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fields::qm31::QM31;
+use thiserror::Error;
+
+use super::memory::{MemoryCache, MemoryEntry, TraceEntry};
+
+#[derive(Debug, Error)]
+pub enum InstructionError {
+    #[error("Unexpected end of memory iterator while processing instruction")]
+    UnexpectedEndOfMemory,
+    #[error("Unknown opcode: {0}")]
+    UnknownOpcode(u32),
+}
+
+/// Opcode wrapper that provides type-safe access to opcode IDs
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Opcode {
+    id: u32,
+}
+
+impl Opcode {
+    /// Get the opcode ID as a u32
+    pub const fn id(self) -> u32 {
+        self.id
+    }
+}
```

Comment:
I think that either we build a full opcode table with all the details (like number of consumed memory slots) we can just drop this for now

Reply:
I agree with the spirit of this idea